homeassistant/components/enphase_envoy/sensor.py

```diff
@@ -65,53 +76,94 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
     envoy_reader = EnvoyReader(ip_address, username, password)
 
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        async with async_timeout.timeout(30):
+            try:
+                data = await envoy_reader.update()
+            except httpcore.ProtocolError as err:
+                raise UpdateFailed(f"Error communicating with API: {err}") from err
+            except httpcore.ConnectTimeout as err:
+                raise UpdateFailed(f"Timeout error with API: {err}") from err
+
+            _LOGGER.debug("Retrieved data from API: %s", data)
+
+            if "can't handle event type ConnectionClosed" in str(
+                data.get("inverters_production")
+            ):
+                raise UpdateFailed("Inverter updated failed")
+
```

Comment:
Can we use a typed exception here instead to avoid the string matching?

Reply:
The api being used is returning a string with the exception inside that. I know not the best solution but will need to be addressed. Without changing the library. I only added these because they were not being caught in the previous sensor code. The api is returning the same as it was previously. I can remove the string checks and go back to the previous behavior of the sensor code which was to really not check much?

Reply:
While it's not ideal, it's preferable for the library to do string matching and throw the typed exception if you don't have another way to handle it.

That way if the api changes we only have to bump the library and don't have to wait for a free reviewer on the HomeAssistant side

Reply:
Let me see what I can come up with in a few days.

Reply:
I have changed the `async_update_data()` method to use try/except blocks rather than string matching.
Take a look at the updated code. I had to change the library to raise exceptions rather than returning the strings.

The api change log is [here](https://github.com/jesserizzo/envoy_reader/compare/master...gtdiehl:exception_handling_ha?expand=1)

I will release the updated library tomorrow and open a separate PR here to bump the `enphase_envoy` library to `0.17.1`